mlir/lib/Transforms/RemoveDeadValues.cpp

```diff
@@ -563,6 +575,48 @@ static void cleanRegionBranchOp(RegionBranchOpInterface regionBranchOp,
   dropUsesAndEraseResults(regionBranchOp.getOperation(), resultsToKeep.flip());
 }
 
+// 1. Iterate over each successor block of the given BranchOpInterface
+// operation.
+// 2. For each successor block:
+//    a. Retrieve the operands passed to the successor.
+//    b. Use the provided liveness analysis (`RunLivenessAnalysis`) to determine
+//    which
+//       operands are live in the successor block.
+//    c. Mark each operand as live or dead based on the analysis.
+// 3. Remove dead operands from the branch operation and arguments accordingly
+
+static void cleanBranchOp(BranchOpInterface branchOp, RunLivenessAnalysis &la) {
```

Comment:
I think this needs to be a recursive solution, you can have the same situation in a nested conditional branch, for example with in a branch you could declare variables and then pass them to the nested conditional branch in this case it won't be part of your successor block arguments.

Reply:
This function is called in a `walk()` which itself will recursively visit the IR.

That said: what about adding a test to cover the case that @codemzs is describing?

Reply:
I added such test, but after looking closely I realized that `walk` and `walk<WalkOrder::PostOrder>` does not help with BranchOp, probably because they do not have parent-child relation. Sorry, I was wrong, it does not reapply iteratively by itself. I am going to add a recursion in some form and let you know.

Reply:
I am quite surprised `walk()` does not traverse nested branch ops, I would also try creating this IR programmatically to ensure you are not missing anything.

Reply:
it does, the issue is the ordering, I need "inner" branches to be traversed first, that is what PostOrder does, but in case of Branches there is not so much hierarchy. I am going to play with it a bit and get back to you